### DISCUSSION TITLE
Pluggable / disabled documentStore

### DISCUSSION THREAD

**Discussion by owenallenaz** [apollographql/apollo-server#5634] [Ideas]
@glasser This originates from https://github.com/apollographql/apollo-server/pull/5629 where there was discussion about making the `documentStore` pluggable. Before I submit a PR I would like to finalize the design of the feature to make sure we're on the same page.

# Current Problem

This is what we're doing in our application now, it's dirty, and reaching into the private variables of `ApolloServer` and I'd rather not, but it's a neccessity because we have 20 ApolloServers that largely share query signatures running in the same Node process.

The `sharedDocumentStore` is initialized early in the App, and then we hit code where we loop and build our 20 ApolloServer's.

```js
const sharedDocumentStore = new InMemoryLRUCache({
	maxSize: Math.pow(2, 20) * 30,
	sizeCalculator: function (obj) {
		return Buffer.byteLength(JSON.stringify(obj), 'utf8');
	}
});
```

```js
for (let schema of schemas) {
	const server = new ApolloServer({
		schema,
		context : ({ req }) => {
			return {
				headers : req.headers
			}
		},
		stopOnTerminationSignals : false
	});

	const schemaDerivedData = await server.schemaDerivedData;
	schemaDerivedData.documentStore = sharedDocumentStore;
}
```

In this case, I'm swapping out the internal private variable so that we can re-use the store for all of our servers. Again, it's dirty, and I'd rather not.

# Proposal

The contract I was looking to change, and seems pretty easy to change in the code is something like:

```js
new ApolloServer({
  ...args,
  documentStore
});
```

With the config argument `documentStore` accepting a more pluggable storage definition in TypeScript like:

```ts
interface PluggableDocumentStore<T> {
    set(key: string, val: T): void
    get(key: string): T
}
```

Then we'd do a small alteration in `generateSchemaDerivedData` where we use the passed documentStore or we initialize the default.

```ts
private generateSchemaDerivedData(schema: GraphQLSchema): SchemaDerivedData {
    const schemaHash = generateSchemaHash(schema!);

    // Use a passed in documentStore or the default
    const documentStore = this.config.documentStore || this.initializeDocumentStore();

    return {
      schema,
      schemaHash,
      documentStore,
    };
  }
```

There is a comment in the code indicates that there was possibly a desire to disable the documentStore `This cannot currently be disabled.` If that's the case, it's possible that we could do `documentStore : PluggableDocumentStore<DocumentNode> | boolean` and if it's false make the code updates needed to effectively skip the pathway. The assumption being that `undefined` or `true` means use the default, `false` means no store, or a passed in value means use the passed in value.

I can submit a new PR with the changes once we decide on the design you think is appropriate.

**Comment by glasser** [comment]
What are the differences between your ApolloServers? Do they share the same schema?

**Reply by owenallenaz** [reply]
We have 20 downstream microservices that each have their own independent graphql server. This application fronts them and exposes them, handles auth, that sort of stuff. This means that we get a system like

```
ApolloServer1{product1_v1, product1_v2, product1_v3, product2_v1, product2_v2}
ApolloServer2{product1_v1}
ApolloServer3{product1_v2}
ApolloServer4{product1_v3}
```

Each is exposed under a different URL, and when the downstream schema changes, we safely hot reload the ApolloServer and pull down the latest schema. Given this architecture there is considerable query overlap for 2 reasons. First v1 and v2 of the same product are almost identical (save for a few breaking differences), also the dedicated server that only servers 1 schema and the main server that serves all schema again will be identical for a specific query.

So a query like 

```
query {
  dms(acct_id: "10") {
    accounts {
      docs {
        acct_id
        label
      }
    }
  }
}
```

That query might be valid on 3 or 4 different servers. When it comes to the persisted query system, this wouldn't be appropriate, but when it comes to the `documentStore` which just handles saving time on re-processing the query, it's re-usable.

**Reply by glasser** [reply]
Sorry, I am not sure if that answers my question: do they have the same schema or different schemas?

The document store doesn't just cache the map from string to parsed document: it also implies that the document is valid for the current schema. So if you share document stores across servers with different schemas, your servers may end up executing invalid operations, causing bugs or even security issues.

**Reply by owenallenaz** [reply]
Your comment, made me worried my implementation so I dug deeper. Let me try and be a little more clear on what we're doing to make sure I'm not about to blow up my production instance. In this case, we have our root `sv-graphql` which is a node process at (some-public-ip) which schema stitches in downstream schemas from private ips, it doesn't have any schemas of it's own.

So the root ApolloServer has the following ApolloServer's inside of it:

ApolloServer1 which uses `/` and stitches it's schemas from `crm-v1` (10.1.1.1), `auth-v1` (10.1.1.2), `auth-v2` (10.1.1.3) and `geo` (10.1.1.4).

ApolloServer2 which uses `/link/geo/` loads it's schemas from `geo` (10.1.1.4).

This means, technically `/` and the `/link/geo/` have some schema portions in common, but are loading different schemas in totality because the ApolloServer1 contains additional schemas that ApolloServer2 does not. In addition, because it's a separate ApolloServer, it's `schema` reference in memory is also a separate `schema` instance.

I undid my hack, to instrument some of the internals using the Node inspector.

I issued a request to `/` (which hits ApolloServer1) with the following query:

```js
query {
  geo {
    ip_to_geo(ip: "35.24.39.129") {
      success
      message
      doc {
        country_code
        country_name
        region_code
        region_name
      }
    }
  }
}
```

The resulting `requestContext.document` was:

```json
{"kind":"Document","definitions":[{"kind":"OperationDefinition","operation":"query","variableDefinitions":[],"directives":[],"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"geo","loc":{"start":10,"end":13}},"arguments":[],"directives":[],"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"ip_to_geo","loc":{"start":20,"end":29}},"arguments":[{"kind":"Argument","name":{"kind":"Name","value":"ip","loc":{"start":30,"end":32}},"value":{"kind":"StringValue","value":"35.24.39.129","block":false,"loc":{"start":34,"end":48}},"loc":{"start":30,"end":48}}],"directives":[],"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"success","loc":{"start":58,"end":65}},"arguments":[],"directives":[],"loc":{"start":58,"end":65}},{"kind":"Field","name":{"kind":"Name","value":"message","loc":{"start":72,"end":79}},"arguments":[],"directives":[],"loc":{"start":72,"end":79}},{"kind":"Field","name":{"kind":"Name","value":"doc","loc":{"start":86,"end":89}},"arguments":[],"directives":[],"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"country_code","loc":{"start":100,"end":112}},"arguments":[],"directives":[],"loc":{"start":100,"end":112}},{"kind":"Field","name":{"kind":"Name","value":"country_name","loc":{"start":121,"end":133}},"arguments":[],"directives":[],"loc":{"start":121,"end":133}},{"kind":"Field","name":{"kind":"Name","value":"region_code","loc":{"start":142,"end":153}},"arguments":[],"directives":[],"loc":{"start":142,"end":153}},{"kind":"Field","name":{"kind":"Name","value":"region_name","loc":{"start":162,"end":173}},"arguments":[],"directives":[],"loc":{"start":162,"end":173}}],"loc":{"start":90,"end":181}},"loc":{"start":86,"end":181}}],"loc":{"start":50,"end":187}},"loc":{"start":20,"end":187}}],"loc":{"start":14,"end":191}},"loc":{"start":10,"end":191}}],"loc":{"start":6,"end":193}},"loc":{"start":0,"end":193}}],"loc":{"start":0,"end":193}}
```

With a md5sum of `ed292e613e8d9887529feee03f258eeb`.

If I issue the SAME request to `/link/geo/` (which hits ApolloServer2). I get an identical `requestContext.document` with the same hash.

Looking at the code, the `graphql.parse` function only takes the arguments `query` and `parseOptions`. In this circumstance `parseOptions` is undefined. I believe `parse()` is idempotent, so it should have no side-effects, and doesn't use any external state outside of the arguments passed in. So then shouldn't the same `query` ALWAYS result in the exact same `requestContext.document`. Basically if I query for `query { foo { bar } } ` that resulting `requestContext.document` should always be the same, regardless of what schema will handle the request. I don't see any code in that pathway that actually interacts with the server's schema, only the schema of the query.

Am I missing something?

**Reply by glasser** [reply]
Yeah, look at the larger context (I've elided some lines relating to calling plugin hooks):

```
  if (config.documentStore) {
    try {
      requestContext.document = await config.documentStore.get(queryHash);
    } catch (err) {
      // ...
    }
  }

  if (!requestContext.document) {
    // ...
    try {
      requestContext.document = parse(query, config.parseOptions);
      // ...
    } catch (syntaxError) {
      // ...
      return await sendErrorResponse(syntaxError, SyntaxError);
    }
    // ...
    const validationErrors = validate(requestContext.document);

    if (validationErrors.length === 0) {
      // ...
    } else {
      // ...
      return await sendErrorResponse(validationErrors, ValidationError);
    }

    if (config.documentStore) {
      Promise.resolve(
        config.documentStore.set(queryHash, requestContext.document),
      ).catch((err) =>
        logger.warn(
          'Could not store validated document. ' + err?.message || err,
        ),
      );
    }
  }
```

If a document is found in the document store, then we skip calling `validate` entirely! (`validate` is a pure function of document plus schema (and not, say, variables or request context).) It's roughly what is responsible for the [validation](http://spec.graphql.org/draft/#sec-Validation) part of the spec, including basic things like "do the queried fields exist" and "do they have the right types for the things you're doing with them syntactically".

So sharing a documentStore across schemas is dangerous. This is why it's on "schemaDerivedData"!

That doesn't mean your overall desire to be able to customize the document store and share the underlying resources is a bad one... just that you'd want to wrap each server object's store in something like PrefixingKeyValueCache.

That said, this also means that the different servers wouldn't be able to share the same cached ASTs, but perhaps your different servers don't actually receive identical documents?

**Reply by owenallenaz** [reply]
Yup, that's a problem. I've verified with my schema that the validation can result in some weird circumstances. Basically if a field is required in one schema, but not another, then the same query string might work in one case but fail on a diff schema. I'll fix that on my end.

Given that issue, I think this idea still has merit. In this case what I'd want to do is pass my own `{ get, set }` which would handle prefixing per downstream schema. This would still allow me to share the LRU cache size across all of the servers, so that way this Node process wouldn't exceed 30mb total cache, instead of the potentiality for 600mb cache.

So the issues with my hackery aside in my personal app, are you OK with the proposal from my original item in terms of the implementation plan?

**Reply by glasser** [reply]
Yes, I think it would be fine, though we would just use KeyValueCache<DocumentNode> rather than a new abstraction — Apollo Server already questionably has one ecosystem of cache implementations, we don't need another.

The docs would have to be *very clear* that you should not actually share a store across servers.

**Reply by owenallenaz** [reply]
Ok, will do. I'll begin work on the PR and then we'll go from there. Thanks.

**Reply by glasser** [reply]
Merged in #5644

**Reply by glasser** [reply]
Unfortunately, if this feature was used in Apollo Gateway (ie, any `new ApolloServer` call receiving both non-null/undefined `documentStore` and `gateway` options), it could lead to a bug where invalid operations could be executed after a schema update. If you are using this feature in a Gateway, we highly recommend upgrading to v3.6.6 which fixes the issue.

The fix is that we now always prefix keys in the document store with a random string that is refreshed each time a new schema is received. We apply this fix in the non-Gateway case too. The fix to the issue also means that the caveat in the docs about not sharing a documentSource across ApolloServer objects has been removed, as this random prefix helps with that case as well.
